evaluation.py

```diff
@@ -14,84 +14,82 @@
 
     black_net: Instance of minigo.DualNetwork, a wrapper around a tensorflow
         convolutional network.
     white_net: Instance of the minigo.DualNetwork.
     games: number of games to play. We play all the games at the same time.
+    sgf_dir: directory to write the sgf results.
     readouts: number of readouts to perform for each step in each game.
     """
 
     # For n games, we create lists of n black and n white players
-    black_players = [MCTSPlayer(
-        black_net, verbosity=verbosity, two_player_mode=True) for i in range(games)]
-    white_players = [MCTSPlayer(
-        white_net, verbosity=verbosity, two_player_mode=True) for i in range(games)]
-
-    # Each player pair represents two players that are going to play a game.
-    player_pairs = [(b, w) for b, w in zip(black_players, white_players)]
+    black = MCTSPlayer(
+        black_net, verbosity=verbosity, two_player_mode=True, num_parallel=SIMULTANEOUS_LEAVES)
+    white = MCTSPlayer(
+        white_net, verbosity=verbosity, two_player_mode=True, num_parallel=SIMULTANEOUS_LEAVES)
 
-    done_pairs = []
+    black_name = os.path.basename(black_net.save_file)
+    white_name = os.path.basename(white_net.save_file)
 
-    # The number of moves that have been played
-    num_moves = 0
+    for i in range(games):
+        num_move = 0  # The move number of the current game
 
-    for black, white in player_pairs:
         black.initialize_game()
         white.initialize_game()
 
-    # The heart of the game-playing loop. Each iteration through the while loop
-    # plays one move for each player. That means we:
-    #   - Do a bunch of MTCS readouts (for each active player, for each game)
-    #   - Play a move (for each active player, for each game)
-    #   - Remove any finished player-pairs
-    while player_pairs:
-        start = time.time()
-
-        for _ in range(readouts):
-            leaves = [pair[num_moves % 2].root.select_leaf()
-                      for pair in player_pairs]
-            probs, vals = (black_net, white_net)[num_moves % 2].run_many(
-                [leaf.position for leaf in leaves])
-
-            for pair, leaf, prob, val in zip(player_pairs, leaves, probs, vals):
-                leaf.incorporate_results(
-                    prob, val, up_to=pair[num_moves % 2].root)
-
-        # print some stats on the search
-        if verbosity >= 3:
-            print(player_pairs[0][0].root.position)
-
-        for black, white in player_pairs:
-            active = white if num_moves % 2 else black
-            inactive = black if num_moves % 2 else white
+        while True:
+            start = time.time()
+            for _ in range(readouts):
+                player = white if num_move % 2 else black
+                player.tree_search()
+
+            # print some stats on the search
+            if verbosity >= 3:
+                print(player.root.position)
+
+            active = white if num_move % 2 else black
```

Comment:
this computation should happen once at top

Reply:
Done